### PR TITLE
Forces docker login on calculate-docker-image and pull-docker-image

### PR DESCRIPTION
**Pull request by jeanschmidt**
Docker login is required if the goal is to instead of relying on EC2 node permission, use OIDC to login instead.

### CODE REVIEW

**Review by zxiiro — APPROVED** [review]

File: .github/actions/calculate-docker-image/action.yml [hunk 1]
```diff
@@ -76,6 +76,26 @@ runs:
...
+        }
+
+        retry () {
+          $*  || (sleep 1 && $*) || (sleep 2 && $*)
+        }
+
+        retry login "${DOCKER_REGISTRY}"
+
     - name: Check if image should be built
       id: check-image
```
Comment (atalman) [inline .github/actions/calculate-docker-image/action.yml:100]: What is the context here ? Could you please add some PR description.

There is an extra if statement in Build and push docker image  step like this, why do we want to skip it ? 
```
if: ${{ steps.calculate-image.outputs.skip != 'true' && (inputs.always-rebuild || steps.check-image.outputs.rebuild) }}
```

**Review by atalman — CHANGES_REQUESTED** [review]

**Comment by atalman** [comment]
if OIDC is required, please use these 2 steps: https://github.com/pytorch/test-infra/blob/atalman-test/.github/workflows/linux_gh_job.yml#L123

**Comment by atalman** [comment]
@jeanschmidt this action is used in multiple core workflows:
```
.github/workflows/_android-build-test.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_android-full-build-test.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_bazel-build-test.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_docs.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_linux-build.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_linux-test.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_rocm-test.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/_xpu-test.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/docker-builds.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
.github/workflows/target-determination-indexer.yml:        uses: pytorch/test-infra/.github/actions/calculate-docker-image@main
```

Lets commit it to a separate branch and create a test PR  in pytorch/pytorch to test the changes.

**Comment by jeanschmidt** [comment]
@atalman tests are being performed in https://github.com/pytorch/pytorch/pull/119544/files 

this is a critical workflow to enable ARC, so I will be following with the less beautiful solution with `docker login` and we can reiterate into figure out why AWS own actions are not working

**Comment by jeanschmidt** [comment]
Relevant signals gathered by running the pull job on pytorch/pytorch

https://github.com/pytorch/pytorch/actions/runs/8237759663/job/22527270971

**Review by DanilBaibak — APPROVED** [review]

**Comment by jeanschmidt** [comment]
Bypassing merge protections, given that signals are green and I believe we're in a good position to move forward.

**Comment by atalman** [comment]
@jeanschmidt 
```
Unable to find image '308535385114.dkr.ecr.us-east-1.amazonaws.com/tool/alpine:latest' locally
docker: Error response from daemon: Head "https://308535385114.dkr.ecr.us-east-1.amazonaws.com/v2/tool/alpine/manifests/latest": no basic auth credentials.
See 'docker run --help'.
```

This is the error I am seeing:
https://github.com/pytorch/pytorch/actions/runs/8251057961/job/22567219192?pr=121728

**Comment by huydhn** [comment]
I happen to have a PR https://github.com/pytorch/pytorch/pull/121691 to switch `308535385114.dkr.ecr.us-east-1.amazonaws.com/tool/alpine` to the upstream image https://hub.docker.com/_/alpine (cause why not) and it seems to work.  May be it would help resolve this error?

**Comment by atalman** [comment]
I think @jeanschmidt added this change to support arc runner. If Arc can support [hub.docker.com/_/alpine](https://hub.docker.com/_/alpine) We probably should be good.

**Comment by huydhn** [comment]
The clear benefit is that we don't need to sync images from docker hub to our ECR anymore.  This is a manual step that I don't think anyone would remember to do from time to time

**Pull request merged**